syntax = "proto3";
package databus;
import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";


// set up connection between user and GM
service Connection {
    rpc Connect (ConnectRequest) returns (ConnectResponse);
}
message ConnectRequest {
    string host = 1;
    string user = 2;
    string password = 3;
    optional int32 timeout = 4;
}
enum ConnectionRC {
    SUCCESSFUL = 0;
    FAILED = 1;
    TIMEOUT = 2;
}
message ConnectResponse {
    ConnectionRC rc = 1;
    string resp = 2;
}


// register a DLT Client (DLTC) to the Global Manager (GM)
service Registration {
    rpc Register (RegisterRequest) returns (RegisterResponse);
    rpc Deresgister (DeregisterRequest) returns (DeregisterResponse);
}
message ResourceInfo { // Pod resource info
    // TODO
}
message RegisterRequest {
    string datasetURL = 1;
    optional bool useCache = 2;
    optional int32 flushFreq = 3;
    optional int32 durability = 4;
    optional int32 logFreq = 5; // frequency of collecting logs
    optional int32 logLinger = 6; // batch log every `logLinger` seconds
    optional int32 logBatchSize = 7; // batch log every `logBatchSize` items
}
message RegisterResponse {
    oneof response {
        RegisterSuccess regsucc = 1;
        RegisterError regerr = 2;
    }
}
message Policy {
    int32 chunkSize = 1;
    repeated string chunkKeys = 2; // an array of chunk keys
}
message RegisterSuccess {
    string jobID = 1;
    string token = 2;
    Policy policy = 3;
    google.protobuf.Timestamp tokenTimeout = 4;
}
message RegisterError {
    string error = 1;
}

message DeregisterRequest {
    string jobID = 1;
    string token = 2;
    optional bool deleteDataset = 3;
}

message DeregisterResponse {
    string response = 1;
}

service UpdatePolicy {
    rpc Update (UpdatePolicyRequest) returns (google.protobuf.Empty) {};
}
message UpdatePolicyRequest {
    string jobID = 1;
    string token = 2;
    Policy policy = 3;
}


service Logger {
    rpc call (google.protobuf.Empty) returns (stream LogItem) {};
}
enum LogLevel {
    INFO = 0;
    DEBUG = 1;
    ERROR = 2;
}
message LogItem {
    google.protobuf.Timestamp time = 1;
    string jobID = 2;
    LogLevel level = 3;
    int32 epoch = 4;
    string details = 5;
}